core/src/main/scala/kafka/common/InterBrokerSendThread.scala

```diff
@@ -84,9 +91,112 @@ abstract class InterBrokerSendThread(name: String,
     }
   }
 
-  def wakeup(): Unit = networkClient.wakeup()
+  private def checkDisconnects(now: Long): Unit = {
+    // any disconnects affecting requests that have already been transmitted will be handled
+    // by NetworkClient, so we just need to check whether connections for any of the unsent
+    // requests have been disconnected; if they have, then we complete the corresponding future
+    // and set the disconnect flag in the ClientResponse
+    for (node <- unsentRequests.nodes) {
+      if (networkClient.connectionFailed(node)) {
+        // Remove entry before invoking request callback to avoid callbacks handling
+        // coordinator failures traversing the unsent list again.
+        val requests = unsentRequests.remove(node)
+        for (request: ClientRequest <- requests) {
+          val handler = request.callback
+          val authenticationException = networkClient.authenticationException(node)
+          if (authenticationException == null) {
+            handler.onComplete(new ClientResponse(request.makeHeader(request.requestBuilder().latestAllowedVersion()),
+              handler, request.destination, now /* createdTimeMs */ , now /* receivedTimeMs */ , true /* disconnected */ ,
+              null /* versionMismatch */ , null /* responseBody */))
+          }
+        }
+      }
+    }
+  }
 
+  private def failExpiredRequests(now: Long): Unit = {
+    // clear all expired unsent requests
+    val expiredRequests = unsentRequests.removeExpiredRequests(now, unsentExpiryMs)
+    for (request: ClientRequest <- expiredRequests)
+      info(s"Failed to send the following request after $unsentExpiryMs ms: ${request.toString}")
+  }
+
+  def wakeup(): Unit = networkClient.wakeup()
 }
 
 case class RequestAndCompletionHandler(destination: Node, request: AbstractRequest.Builder[_ <: AbstractRequest],
-                                       handler: RequestCompletionHandler)
\ No newline at end of file
+                                       handler: RequestCompletionHandler)
+
+class UnsentRequests {
+  private var unsent = new ConcurrentHashMap[Node, ConcurrentLinkedQueue[ClientRequest]]
```

Comment:
Seems we don't really need these to be concurrent collections since they are only accessed by the send thread. Similarly, I don't think we need the synchronization below.